### PR TITLE
Task DEV 4 Part 3

### PR DESCRIPTION
**Pull request by PolinaYeroshevich**
Create an abstract class Figure. Which will have:
1.name field
2.color field
3.constructor that accepts the color and name of the shape as input
4.Abstract CalculatePerimeter method
5.Abstract CalculateArea method
6.Abstract method CheckForExistance
7.Virtual method GetFullInfo, which will return a string containing information about the name and color of the shape
Create a descendant class of Triangle (inherit from a shape)
1.baseSide field
2.firstSide field
3. secondSide field
4. Implement the void CheckForExistance method
a. which will check if such a triangle can be created (A triangle exists only when the sum of its two sides is greater than the third). If the triangle does not exist, throw an exception with the message why the triangle could not be created
b. which checks that the sides of the triangle are not negative and greater than zero, otherwise we throw an exception with a message why the triangle cannot be created
5. a constructor that sets all information about a triangle (sides, color, name of the figure), if such a triangle exists (to check the existence of a triangle, the method from p. 4 is used).
6.implement the CalculatePerimeter method
7.implement the CalculateArea method
8.override the GetFullInfo method, which will return a string containing information about the title, color and side lengths
Create a Parallelogram inheritor class (inherit from a figure)
1.firstSide field
2. secondSide field
3.field height
4.field angle
5. Implement the void CheckForExistance method
6. Two constructors as in the previous task using the existence test
9.implement the CalculatePerimeter method
10.implement the CalculateArea method
11.override the GetFullInfo method, which will return a string containing all information about the parallelogram
Create a class inheriting from Circle (inherit from a shape)
• Filling by analogy with triangle and parallelogram
Create a class inheriting from Square (inherit from a shape)
• Filling by analogy with triangle and parallelogram

### CODE REVIEW

File: Task_DEV_4/Task_DEV_4/Figura.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿namespace Task_DEV_4
+{
+    public abstract class Figura
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Figura.cs]: FigurE

File: Task_DEV_4/Task_DEV_4/Figura.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public string Name { get; set; }
+        public string Color { get; set; }
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Figura.cs]: Alphabet order.

File: Task_DEV_4/Task_DEV_4/Figura.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        public abstract void CheckForExistance();
+
+        public abstract double CalculatePerimeter();
+
+        public abstract double CalculateArea();
+
+        public virtual string GetFullInfo()
+        {
+            return $"{Name},{Color}";
+        }
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Figura.cs]: Alphabet order.

File: Task_DEV_4/Task_DEV_4/Figura.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+        public abstract void CheckForExistance();
+
+        public abstract double CalculatePerimeter();
+
+        public abstract double CalculateArea();
+
+        public virtual string GetFullInfo()
+        {
+            return $"{Name},{Color}";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Figura.cs]: Space after coma.

File: Task_DEV_4/Task_DEV_4/Circle.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Circle : Figura
+    {
+        public double Radius { get; set; }
+        private const double PI = 3.14;
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Circle.cs]: Why do you create this constant if it already exists in Math? Math.PI

File: Task_DEV_4/Task_DEV_4/Circle.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+        public override void CheckForExistance()
+        {
+            if (Radius <= 0)
+            {
+                throw new ArgumentOutOfRangeException("Side cannot be less than 0.");
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Circle.cs]: Side or radius?

File: Task_DEV_4/Task_DEV_4/Circle.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+            }
+        }
+
+        public Circle(double radius, string name, string color) : base(name, color)
+        {
+            if (Radius > 0)
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Circle.cs]: Why this check if you have CheckForExistance and can use it here?

**Review by VadimBelkin — CHANGES_REQUESTED** [review]

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs [hunk 8]
```diff
@@ -8,21 +8,37 @@ static void Main(string[] args)
...
+                Console.WriteLine(fullInfoSquare);
 
-                Rectangle rectangle = new Rectangle(4, 7);
-                double rectangleArea = rectangle.CalculateArea();
-                Console.WriteLine($"The area of a rectangle is {rectangleArea}");
-                double rectanglePerimeter = rectangle.CalculatePerimeter();
-                Console.WriteLine($"The perimeter of a rectangle is {rectanglePerimeter}");
-
-                Parallelogram parallelogram = new Parallelogram(2, 3);
+                Parallelogram parallelogram = new Parallelogram(9, 7, 5, "blue", "Parallelogram");
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/EntryPoint.cs]: Is the order of the variables correct?

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs [hunk 9]
```diff
@@ -8,21 +8,37 @@ static void Main(string[] args)
...
                 double parallelogramArea = parallelogram.CalculateArea();
                 Console.WriteLine($"The area of a paralelogram is {parallelogramArea}");
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/EntryPoint.cs]: typo

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs [hunk 10]
```diff
@@ -8,21 +8,37 @@ static void Main(string[] args)
...
+                double parallelogramPerimeter = parallelogram.CalculatePerimeter();
+                Console.WriteLine($"The perimeter of a parallelogram is {parallelogramPerimeter}");
+                String fullInfoParallelogram = parallelogram.GetFullInfo();
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/EntryPoint.cs]: write string with a lower case

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 11]
```diff
@@ -2,37 +2,75 @@
...
+                throw new ArgumentOutOfRangeException("Side cannot be less than 0.");
+            }
+            else if (FirstSide <= 0 || SecondSide <= 0 || Angle <= 0)
+            {
+                throw new ArgumentOutOfRangeException("Side cannot be less than 0.");
+            }
         }
 
-        public Parallelogram(double sideA, double angle, double sideB) : base(sideA)
+        public Parallelogram(double firstSide, double height, string name, string color) : base(color, name)
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: wrong order of the variables

File: Task_DEV_4/Task_DEV_4/Figura.cs (hunk 1 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Figura.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Figura.cs (hunk 2 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Figura.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Figura.cs (hunk 3 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Figura.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Figura.cs (hunk 4 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Figura.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Circle.cs (hunk 5 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Circle.cs (hunk 6 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Circle.cs (hunk 7 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Figura.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
+﻿namespace Task_DEV_4
+{
+    public abstract class Figure
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Figura.cs]: Class name was changed but in the name of file typo remains.

File: Task_DEV_4/Task_DEV_4/Circle.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Circle : Figure
+    {
+        private const string errorMessage = "Radius cannot be less than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Circle.cs]: ERROR_MESSAGE

Comment (VadimBelkin) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Need to insert one blank line between fields and properties.

Comment (VadimBelkin) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: less or equal

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 14]
```diff
@@ -2,37 +2,55 @@
 
 namespace Task_DEV_4
 {
-    class Parallelogram : Square
+    class Parallelogram : Figure
     {
-        public double HeightA { get; set; }
+        private const string errorMessage = "Side cannot be less than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: The same as in comments above.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 15]
```diff
@@ -2,37 +2,55 @@
...
+                return (FirstSide + SecondSide) * 2;
             }
             else
             {
-                throw new ArgumentOutOfRangeException("Side cannot be less than 0");
+                throw new ArgumentOutOfRangeException("Side cannot be less than 0.");
             }
         }
+
+        public override void CheckForExistance()
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Where do you use this method?

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 14 above)
Comment (VadimBelkin) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Only side? What about angle?

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 16]
```diff
@@ -2,37 +2,55 @@
...
+        {
+            if (FirstSide <= 0 || SecondSide <= 0 || Angle <= 0)
+            {
+                throw new ArgumentOutOfRangeException("Side cannot be less than 0.");
+            }
+        }
+
+        public override string GetFullInfo()
+        {
+            return $"Name : {Name}, Color : {Color}, Length of sides : {FirstSide} , {SecondSide}";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Angle?

File: Task_DEV_4/Task_DEV_4/Square.cs [hunk 17]
```diff
@@ -2,23 +2,38 @@
 
 namespace Task_DEV_4
 {
-    class Square
+    class Square : Figure
     {
-        public double SideA { get; set; }
+        private const string errorMessage = "Side cannot be less than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Square.cs]: The same as in comments above.

File: Task_DEV_4/Task_DEV_4/Square.cs [hunk 18]
```diff
@@ -2,23 +2,38 @@
...
+        {
+            if (Side <= 0)
+            {
+                throw new ArgumentOutOfRangeException(errorMessage);
+            }
+        }
+
+        public override string GetFullInfo()
+        {
+            return $"Name : {Name}, Color : {Color},Length of side : {Side}";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Square.cs]: Space after coma

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 19]
```diff
@@ -0,0 +1,49 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Triangle : Figure
+    {
+        private const string errorMessage = "Side cannot be less than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: The same as in comments above.

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+            FirstSide = firstSide;
+            SecondSide = secondSide;
+            CheckForExistance();
+        }
+
+        public override double CalculateArea()
+        {
+            double perimeter = BaseSide + FirstSide + SecondSide;
+            double semiPerimeter = perimeter / 2;
+            return Math.Sqrt(semiPerimeter * (semiPerimeter - BaseSide) * (semiPerimeter - FirstSide) * (semiPerimeter - SecondSide));
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: Too long string

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+        public override double CalculatePerimeter()
+        {
+            return BaseSide + FirstSide + SecondSide;
+        }
+
+        public override void CheckForExistance()
+        {
+            if (BaseSide + FirstSide < SecondSide || BaseSide + SecondSide < FirstSide || SecondSide + FirstSide < BaseSide)
+            {
+                throw new ArgumentOutOfRangeException("The sum of the two sides cannot be less than the third side.");
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: Message to const

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        public override double CalculatePerimeter()
+        {
+            return BaseSide + FirstSide + SecondSide;
+        }
+
+        public override void CheckForExistance()
+        {
+            if (BaseSide + FirstSide < SecondSide || BaseSide + SecondSide < FirstSide || SecondSide + FirstSide < BaseSide)
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: Too long string.

**Review by VadimBelkin — CHANGES_REQUESTED** [review]

File: Task_DEV_4/Task_DEV_4/Figura.cs (hunk 12 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Figura.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Circle.cs (hunk 13 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 14 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 15 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 16 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Square.cs (hunk 17 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Square.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Square.cs (hunk 18 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Square.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 19 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 20 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 21 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 22 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Circle.cs [hunk 23]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Circle : Figure
+    {
+        private const string ERROR_MESSAGE = "Radius cannot be less equals than 0.";
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Circle.cs]: Is it correct message?

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 24]
```diff
@@ -2,37 +2,57 @@
 
 namespace Task_DEV_4
 {
-    class Parallelogram : Square
+    class Parallelogram : Figure
     {
-        public double HeightA { get; set; }
-        public double Angle { get; set; }
-        public double SideB { get; set; }
+        private const string ERROR_MESSAGE = "Side or angle cannot be less equals than 0.";
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Is it correct message?

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 25]
```diff
@@ -2,37 +2,57 @@
...
+            SecondSide = secondSide;
+            FirstSide = firstSide;
             Angle = angle;
+            CheckForExistance();
         }
 
         public override double CalculateArea()
         {
-            if (HeightA > 0 && SideA > 0)
+            if (FirstSide > 0 && SecondSide > 0 && Angle > 0)
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Why are you using this check in this method?

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 26]
```diff
@@ -2,37 +2,57 @@
...
-            else if (SideA > 0 && SideB > 0 && Angle > 0)
+            else
+            {
+                throw new ArgumentOutOfRangeException(ERROR_MESSAGE);
+            }
+        }
+
+        public override double CalculatePerimeter()
+        {
+            if (FirstSide > 0 && SecondSide > 0)
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Why are you using this check in this method?

File: Task_DEV_4/Task_DEV_4/Square.cs [hunk 27]
```diff
@@ -2,23 +2,39 @@
 
 namespace Task_DEV_4
 {
-    class Square
+    class Square : Figure
     {
-        public double SideA { get; set; }
+        private const string ERROR_MESSAGE = "Side cannot be less equals than 0.";
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Square.cs]: same comment as previous

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 28]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Triangle : Figure
+    {
+        private const string ERROR_MESSAGE = "Side cannot be less equals than 0.";
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: same comment

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            BaseSide = baseSide;
+            FirstSide = firstSide;
+            SecondSide = secondSide;
+            CheckForExistance();
+        }
+
+        public override double CalculateArea()
+        {
+            double perimeter = BaseSide + FirstSide + SecondSide;
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: you can reuse CalculatePerimeter() method for perimeter variable.

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+            return Math.Sqrt(semiPerimeter * (semiPerimeter - BaseSide) *
+                (semiPerimeter - FirstSide) * (semiPerimeter - SecondSide));
+        }
+
+        public override double CalculatePerimeter()
+        {
+            return BaseSide + FirstSide + SecondSide;
+        }
+
+        public override void CheckForExistance()
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: which came first, the chicken or the egg?:) (think about order of checks)

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+            }
+            else if (BaseSide <= 0 || FirstSide <= 0 || SecondSide <= 0)
+            {
+                throw new ArgumentOutOfRangeException(ERROR_MESSAGE);
+            }
+        }
+
+        public override string GetFullInfo()
+        {
+            return $"Name : {Name}, Color : {Color}, Length of sides : {BaseSide}, {FirstSide}, {SecondSide}";
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: wrong code style.

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 32]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Triangle : Figure
+    {
+        private const string ERROR_MESSAGE = "Side cannot be less equals than 0.";
+        private const string ERROR_MESSAGE2 = "The sum of the two sides cannot be less than the third side.";
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: please, give more clarity names for constants.
code style

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs [hunk 33]
```diff
@@ -8,21 +8,37 @@ static void Main(string[] args)
         {
             try
             {
-                Square square = new Square(5);
+                Triangle triangle = new Triangle("Green", "Triangle", 3, 4, 5);
+                double triangleArea = triangle.CalculateArea();
+                Console.WriteLine($"The area of a triangle is {triangleArea}");
+                double trianglePerimeter = triangle.CalculatePerimeter();
+                Console.WriteLine($"The perimeter of a triangle is {trianglePerimeter}");
+                String fullInfoTriangle = triangle.GetFullInfo();
```
Comment (MariaEroshevich) [inline Task_DEV_4/Task_DEV_4/EntryPoint.cs]: lower case

**Review by MariaEroshevich — CHANGES_REQUESTED** [review]

File: Task_DEV_4/Task_DEV_4/Circle.cs (hunk 23 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 24 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 25 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 26 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Square.cs (hunk 27 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Square.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 28 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 29 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 30 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 31 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 32 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs (hunk 33 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/EntryPoint.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 11 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs (hunk 10 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/EntryPoint.cs]: Done.

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs (hunk 9 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/EntryPoint.cs]: Done.

File: Task_DEV_4/Task_DEV_4/EntryPoint.cs (hunk 8 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/EntryPoint.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Circle.cs [hunk 34]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Circle : Figure
+    {
+        private const string MESSAGE_ERROR_FIRST = "Radius cannot be less or equals than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Circle.cs]: Bad variable name. Name it smth like INVALID_RANGE_ERROR_MESSAGE. Name has to reflect sense.

Comment (VadimBelkin) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Less or equal. Without s.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs [hunk 35]
```diff
@@ -2,37 +2,45 @@
 
 namespace Task_DEV_4
 {
-    class Parallelogram : Square
+    class Parallelogram : Figure
     {
-        public double HeightA { get; set; }
+        private const string MESSAGE_ERROR_FIRST = "Side or angle cannot be less or equals than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Parallelogram.cs]: The same as in comments above.

File: Task_DEV_4/Task_DEV_4/Square.cs [hunk 36]
```diff
@@ -2,23 +2,39 @@
 
 namespace Task_DEV_4
 {
-    class Square
+    class Square : Figure
     {
-        public double SideA { get; set; }
+        private const string MESSAGE_ERROR_FIRST = "Side cannot be less or equals than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Square.cs]: The same as in comments above.

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 37]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Triangle : Figure
+    {
+        private const string MESSAGE_ERROR_FIRST = "Side cannot be less or equals than 0.";
+        private const string MESSAGE_ERROR_SECOND = "The sum of the two sides " +
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: The same as in comments above.

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 38]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Triangle : Figure
+    {
+        private const string MESSAGE_ERROR_FIRST = "Side cannot be less or equals than 0.";
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: Bad variable name.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 37 above)
Comment (VadimBelkin) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: smth like TRIANGLE_EXISTANCE_ERROR_MESSAGE

**Review by VadimBelkin — CHANGES_REQUESTED** [review]

File: Task_DEV_4/Task_DEV_4/Circle.cs (hunk 34 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Circle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Parallelogram.cs (hunk 35 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Parallelogram.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Square.cs (hunk 36 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Square.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 37 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 38 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

File: Task_DEV_4/Task_DEV_4/Triangle.cs [hunk 39]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+
+namespace Task_DEV_4
+{
+    class Triangle : Figure
+    {
+        private const string INVALID_RANGE_ERROR_MESSAGE = "Side cannot be less or equal than 0.";
+        private const string TRIANGLE_EXISTANCE_ERROR_MESSAGE = "The sum of the two sides " +
```
Comment (VadimBelkin) [inline Task_DEV_4/Task_DEV_4/Triangle.cs]: existEnce

**Review by VadimBelkin — CHANGES_REQUESTED** [review]

File: Task_DEV_4/Task_DEV_4/Triangle.cs (hunk 39 above)
Comment (PolinaYeroshevich) [reply Task_DEV_4/Task_DEV_4/Triangle.cs]: Done.

**Review by VadimBelkin — APPROVED** [review]

**Review by MariaEroshevich — APPROVED** [review]

**Pull request merged**